api/valor_api/backend/core/llm_clients.py

```diff
@@ -0,0 +1,659 @@
+**
+
+Input:
+{query}
+
+Statements:
+{statements}
+
+JSON:
+"""
+
+
+def _get_coherence_instruction(text: str) -> str:
+    """
+    This instruction was adapted from appendix A of DeepEval's paper G-EVAL: NLG Evaluation using GPT-4 with Better Human Alignment (https://arxiv.org/pdf/2303.16634).
+    The main adaptation is a generalization of the metric to more task types. The example prompt in DeepEval was specific to summarization, but the below prompt could apply to any text generation task.
+    Crucially, unlike DeepEval, no context is used. Instead, the coherence of the text is evaluated entirely based on the text. This generalizes the prompt and also prevents the evaluation from being influenced by the quality of sentences in the context.
+    Parameters
+    ----------
+    text: str
+        The text to be evaluated.
+    Returns
+    -------
+    str
+        The instruction for the llm.
+    """
+    return f"""
+    You are a helpful assistant. You will grade the text. Your task is to rate the text based on its coherence. Please make sure you read and understand these instructions carefully. Please keep this document open while reviewing, and refer to it as needed.
+
+    Evaluation Criteria:
+    Coherence (1-5) - the collective quality of all sentences. We align this dimension with the DUC quality question of structure and coherence whereby ”the summary should be well-structured and well-organized. The summary should not just be a heap of related information, but should build from sentence to sentence to a coherent body of information about a topic.”
+
+    Evaluation Steps:
+    1. Read the text carefully and identify the main topic and key points.
+    2. Check if the text presents the information in a clear and logical order. Examine the collective quality of all sentences.
+    3. Assign a score for coherence on a scale of 1 to 5, where 1 is the lowest and 5 is the highest based on the Evaluation Criteria. Respond with just the number 1 to 5.
+
+    Text to Evaluate:
+    {text}
+
+    Coherence Score (1-5):
+    """
+
+
+class LLMClient:
+    """
+    Parent class for all LLM clients.
+
+    Attributes
+    ----------
+    api_key : str, optional
+        The API key to use.
+    model_name : str
+        The model to use.
+    """
+
+    api_key: str | None = None
+    model_name: str
+
+    def __init__(
+        self,
+        api_key: str | None = None,
+        model_name: str | None = None,
+    ):
+        """
+        Set the API key and model name (if provided).
+        """
+        self.api_key = api_key
+        if model_name is not None:
+            self.model_name = model_name
+
+    def connect(
+        self,
+    ):
+        """
+        Setup the connection to the API. Not implemented for parent class.
+        """
+        raise NotImplementedError
+
+    def process_messages(
+        self,
+        messages: list[dict[str, str]],
+    ) -> Any:
```

Comment:
should this really be Any?

Reply:
Second this, the mistral client below seems to return `list[ChatMessage]` while the OpenAI client uses this base function.

I think you can retain the `-> Any` if you make this base function `raise NotImplementedError` and move the contents to the OpenAI client.

If multiple clients use the default function and you want to preserve a single point of truth then maybe a middle class would be better. The inheritance would be `LLMClient --> OpenAIStyleClient  --> OpenAIClient`